src/libraries/System.Net.Http/src/System/Net/Http/Headers/HttpHeaders.cs

```diff
@@ -864,6 +812,10 @@ private static bool TryParseAndAddRawHeaderValue(HeaderDescriptor descriptor, He
                     {
                         AddParsedValue(info, parsedValue);
                     }
+                    else if (addWhenInvalid && string.IsNullOrWhiteSpace(value))
```

Comment:
I know this is from my change, but I don't think `string.IsNullOrWhiteSpace` is the right condition here.

As far as I know, the only parser that will ever return `true` and set the `parsedValue` to non-null for non-empty inputs is `Cache-Control`.

This check may have to be `else if (addWhenInvalid && !descriptor.Equals(KnownHeaders.CacheControl))`.

(same for the other check below in this method)
I wouldn't be surprised if we're lacking test coverage around these edge cases.

Reply:
> As far as I know, the only parser that will ever return `true` and set the `parsedValue` to non-null for non-empty inputs is `Cache-Control`.

I am interested in looking into improving test coverage. However, could you elaborate slightly more on this, as it sounds like expected behavior unless I am missing something. 
If the parser returns `true`, we should expect `parsedValue` to be non-null if the non-null input is valid.

Reply:
`TryParseValue` returning true means that it consumed some input and considered it valid. It indicates that we shouldn't throw an exception for invalid input.

If it set `parsedValue` to something non-null, it means the parser extracted a single value.
If it did not, the consumed characters are _likely_ just whitespace (which is valid, just not a meaningful parsed value entry). We'll loop again and try to consume the actual value in the next call.

We must always have at least one entry though. If the consumed characters included a separator (e.g. " , , , "), we still want to add the invalid entry.

The logic here should then be
```c#
if (parsedValue is not null)
{
    AddParsedValue(parsedValue);
}
else if (addWhenInvalid)
{
    AddInvalidValue(value ?? string.Empty);
}
```

The only exception is `Cache-Control`. It uses a single parsed value entry to store all parsed values. As such, it will only ever set `parsedValue` once, even as it consumed valid values.

The problem arises if there are multiple Cache-Control headers added. The `parsedValue` will already be set, so subsequent parsing attempts will always see `parsedValue is null`.

---

For this PR, I think we can do something like
```c#
else if (addWhenInvalid && (!descriptor.Equals(KnownHeaders.CacheControl) || info.ParsedAndInvalidValues is null))
```

As a separate PR, we should look at changing the `Cache-Control` parser to behave like all the other parsers and update this check to just
```c#
else if (addWhenInvalid)
```

Reply:
Actually given that parsing can already modify the values somewhat, we should make it just
```c#
else if (addWhenInvalid && info.ParsedAndInvalidValues is null)
```

Reply:
Thanks for clarifying the behavior of the parser, it's quite clear now.
> `else if (addWhenInvalid && info.ParsedAndInvalidValues is null)`

With this change, just `TryGetValues_GetValuesForExistingHeader_ReturnsTrueAndListOfValues` is failing.
I will have some time to dig further into it this weekend, however worst case we can fall back to explicitly checking for CacheControl.

Reply:
Do you mean this new assert?
https://github.com/dotnet/runtime/blob/10751aa74568022778e97fcaebb06c156aed10b3/src/libraries/System.Net.Http/tests/UnitTests/Headers/HttpHeadersTest.cs#L1193

That's fine - we weren't keeping all empty values before either.

Reply:
Sounds good, I've update that. I've also added a test to cover the current edge case: 
https://github.com/dotnet/runtime/pull/67833/files#diff-1926155198b3dbd5ccde8ada0ade66e5b3d816d0421070e08355599cd0604aa1R658 

https://github.com/heathbm/runtime/blob/8607b485f08c3aa4ed3ac3e5556fa80aa14a76fb/src/libraries/System.Net.Http/tests/UnitTests/Headers/CacheControlHeaderValueTest.cs#L658